openmc/deplete/openmc_operator.py

```diff
@@ -0,0 +1,567 @@
+            for mat in self.materials:
+                if str(mat.id) in local_mats:
+                    self._set_number_from_mat(mat)
+
+        # Else from previous depletion results
+        else:
+            for mat in self.materials:
+                if str(mat.id) in local_mats:
+                    self._set_number_from_results(mat, prev_res)
+
+    def _set_number_from_mat(self, mat):
+        """Extracts material and number densities from openmc.Material
+
+        Parameters
+        ----------
+        mat : openmc.Material
+            The material to read from
+
+        """
+        mat_id = str(mat.id)
+
+        for nuclide, atom_per_bcm in mat.get_nuclide_atom_densities().items():
+            atom_per_cc = atom_per_bcm * 1.0e24
+            self.number.set_atom_density(mat_id, nuclide, atom_per_cc)
+
+    def _set_number_from_results(self, mat, prev_res):
+        """Extracts material nuclides and number densities.
+
+        If the nuclide concentration's evolution is tracked, the densities come
+        from depletion results. Else, densities are extracted from the geometry
+        in the summary.
+
+        Parameters
+        ----------
+        mat : openmc.Material
+            The material to read from
+        prev_res : Results
+            Results from a previous depletion calculation
+
+        """
+        mat_id = str(mat.id)
+
+        # Get nuclide lists from geometry and depletion results
+        depl_nuc = prev_res[-1].nuc_to_ind
+        geom_nuc_densities = mat.get_nuclide_atom_densities()
+
+        # Merge lists of nuclides, with the same order for every calculation
+        geom_nuc_densities.update(depl_nuc)
+
+        for nuclide, atom_per_bcm in geom_nuc_densities.items():
+            if nuclide in depl_nuc:
+                concentration = prev_res.get_atoms(mat_id, nuclide)[1][-1]
+                volume = prev_res[-1].volume[mat_id]
+                atom_per_cc = concentration / volume
+            else:
+                atom_per_cc = atom_per_bcm * 1.0e24
+
+            self.number.set_atom_density(mat_id, nuclide, atom_per_cc)
+
+    @abstractmethod
+    def _get_helper_classes(self, helper_kwargs):
+        """Create the ``_rate_helper``, ``_normalization_helper``, and
+        ``_yield_helper`` objects.
+
+        Parameters
+        ----------
+        reaction_rate_mode : str
+            Indicates the subclass of :class:`ReactionRateHelper` to
+            instantiate.
+        normalization_mode : str
+            Indicates the subclass of :class:`NormalizationHelper` to
+            instantiate.
+        fission_yield_mode : str
+            Indicates the subclass of :class:`FissionYieldHelper` to instatiate.
+        reaction_rate_opts : dict
+            Keyword arguments that are passed to the :class:`ReactionRateHelper`
+            subclass.
+        fission_yield_opts : dict
+            Keyword arguments that are passed to the :class:`FissionYieldHelper`
+            subclass.
```

Comment:
This doesn't match what arguments are passed (only `helper_kwargs`)

Reply:
I've removed this. It was cruft which I forgot to move.